lib/CL/devices/level0/level0-driver.cc

```diff
@@ -1871,18 +1978,153 @@ Level0Device::Level0Device(Level0Driver *Drv, ze_device_handle_t DeviceH,
   SStream.flush();
   KernelCacheHash = SStream.str();
 
+  initBuiltinKernels();
+
   ClDev->available = CL_TRUE;
 }
 
 Level0Device::~Level0Device() {
+  destroyBuiltinKernels();
   // ComputeQueues.wait()
   // CopyQueues.wait()
   freeMem(static_cast<void *>(ComputeTimestamps));
   freeMem(static_cast<void *>(CopyTimestamps));
 }
 
+static void calculateHash(uint8_t *BuildHash,
+                          const uint8_t *Data,
+                          const size_t Len) {
+  SHA1_CTX HashCtx;
+  pocl_SHA1_Init(&HashCtx);
+  pocl_SHA1_Update(&HashCtx, Data, Len);
+  pocl_SHA1_Final(&HashCtx, BuildHash);
+  uint8_t TempDigest[SHA1_DIGEST_SIZE];
+  pocl_SHA1_Final(&HashCtx, TempDigest);
+
+  uint8_t *hashstr = BuildHash;
+  for (unsigned i=0; i < SHA1_DIGEST_SIZE; i++)
+      {
+          *hashstr++ = (TempDigest[i] & 0x0F) + 65;
+          *hashstr++ = ((TempDigest[i] & 0xF0) >> 4) + 65;
+      }
+  *hashstr = 0;
+  BuildHash[2] = '/';
+}
+
+bool Level0Device::initBuiltinKernels() {
```

Comment:
Better rename to "HelperKernels" or similar as I started to think this is exposin a set of OpenCL 1.2 BiKs.

Reply:
fixed